src/globus_cli/termio/formatters/base.py

```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
+
+import abc
+import typing as t
+import warnings
+
+T = t.TypeVar("T")
+
+
+class FormattingFailedWarning(UserWarning):
+    pass
+
+
+class FieldFormatter(abc.ABC, t.Generic[T]):
+    parse_null_values: bool = False
+
+    def warn_formatting_failed(self, value: t.Any) -> None:
+        warnings.warn(
+            f"Formatting failed for '{value!r}' with formatter={self!r}",
+            FormattingFailedWarning,
+        )
+
+    @abc.abstractmethod
+    def parse(self, value: t.Any) -> T:
+        """
+        The `parse()` step is responsible for producing well-formed data for a field.
+        For example, parsing may convert a dictionary or mapping to a tuple by pulling
+        out the relevant fields and checking their types.
+
+        If parsing fails for any reason, it should raise a ValueError.
+        """
+
+    @abc.abstractmethod
+    def render(self, value: T) -> str:
+        """
+        The `render()` step is responsible taking data which has already been parsed
+        and reshaped and converting it into a string.
+        For example, rendering may comma-join an array of strings into a
+        comma-delimited list.
+
+        If rendering fails for any reason, it should raise a ValueError.
+        """
+
+    def format(self, value: t.Any) -> str:
+        """
+        Formatting data consists primarily of parsing and then rendering.
+        If either step fails, the default behavior warns and falls back on str().
+        """
+        if value is None and self.parse_null_values is False:
+            return "None"
+        try:
+            data = self.parse(value)
+            return self.render(data)
```

Comment:
Is there a test that exercises raising `ValueError` from `.render()`, specifically when calling this `.format()` method?

If not, adding a test will help pin this call to `.render()` inside the `try` block.

Reply:
Yes, I think this is a good thing to test in the unit tests to ensure we maintain this invariant/constraint/behavior.

Reply:
https://github.com/globus/globus-cli/pull/692/commits/7ad433399ed341e740c219cb32723e7038e86d93 contains unit tests for the base formatter and I think it covers everything which is reasonable to test/assert about the base class. Definitely let me know if you see anything else which we should be checking in those tests.